data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,670 @@
+        unsafe { from_raw_parts(self.ptr, self.len) }
+    }
+
+    #[inline]
+    fn layout(len: usize) -> Layout {
+        if let Ok(layout) = Layout::array::<u8>(len) {
+            layout
+        } else {
+            Self::overflow()
+        }
+    }
+
+    /// There is little that can be done if an allocation fails
+    /// due to incorrect layout or memory exhaustion.
+    #[allow(clippy::panic)]
+    #[inline]
+    const fn overflow() -> ! {
+        panic!("overflow");
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    #[allow(unsafe_code)]
+    fn clone(&self) -> Self {
+        let len = self.len;
+        let layout = Self::layout(len);
+        // SAFETY: pointer is checked to be non-null.
+        let ptr = unsafe { alloc::alloc::alloc(layout) };
+        if ptr.is_null() {
+            Self::overflow();
+        }
+        // SAFETY: all conditions are met.
+        unsafe { ptr.copy_from_nonoverlapping(self.ptr, self.len) };
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let len = value.len();
+        let boxed = value.as_bytes().to_vec().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl From<String> for BoxedString {
+    #[inline]
+    fn from(value: String) -> Self {
+        let len = value.len();
+        let boxed = value.into_boxed_str();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl Drop for BoxedString {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        // SAFETY:
+        // - using global allocator
+        // - `BoxedString`s have the same layout varied only by it's length
+        unsafe {
+            alloc::alloc::dealloc(self.ptr as *mut u8, Self::layout(self.len));
+        }
+    }
+}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Send for BoxedString {}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Sync for BoxedString {}
+
+#[derive(Clone, Copy)]
+#[repr(C)]
+struct InlinedString {
+    #[cfg(target_endian = "little")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+    /// MSB is always 1 to distinguish inlined variant.  
+    len: u8,
+    #[cfg(target_endian = "big")]
```

Comment:
```suggestion
    payload: [u8; MAX_INLINED_STRING_LEN],
    /// MSB is always 1 to distinguish inlined variant.  
    len: u8,
```

I don't think we need to watch the endianess, just have the layouts of `InlinedString` and `BoxedString` be compatible

Reply:
I think that the endianness check could be useful. I'd keep it until it's proven safe.

Reply:
Actually i think endianness is important here, because it does not affect structs offsets.

Little-endian:

```text
+---------+-------+---------+----------+----------------+
| Bits    | 0..63 | 64..118 | 119..126 | 127            |
+---------+-------+---------+----------+----------------+
| Inlined | payload         | len      | tag (MSB)      |
+---------+-------+---------+----------+----------------+
| Box     | ptr   | len                | tag (MSB)      |
+---------+-------+--------------------+----------------+
```

Big-endian without changing layout (order of bytes for `BoxedString::len` would be reversed):

```text
+---------+-------+----------+-----------+---------+----------+-----------+
| Bits    | 0..63 | 64..70   | 71        | 72..118 | 119..126 | 127       |
+---------+-------+----------+-----------+---------+----------+-----------+
| Inlined | payload                                | len      | tag (MSB) |
+---------+-------+----------+-----------+---------+----------+-----------+
| Box     | ptr   | len...   | tag (MSB) | ...len                         |
+---------+-------+----------+-----------+--------------------------------+
```

Reply:
very good, you're correct, I forgot one is 4bytes and the other just 1byte

Reply:
sorry, my bad, I'll resolve this conversation